sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceRDD.scala

```diff
@@ -51,18 +58,19 @@ class DataSourceRDD[T: ClassTag](
         valuePrepared
       }
 
-      override def next(): T = {
+      override def next(): Any = {
         if (!hasNext) {
           throw new java.util.NoSuchElementException("End of stream")
         }
         valuePrepared = false
         reader.get()
       }
     }
-    new InterruptibleIterator(context, iter)
+    // TODO: get rid of this type hack.
+    new InterruptibleIterator(context, iter.asInstanceOf[Iterator[InternalRow]])
```

Comment:
Why is this necessary? I think the TODO should be handled in this commit and that Spark shouldn't cast RDD[ColumnarBatch] to RDD[InternalRow].

What about having the RDD iterate over the rows in the batch to actually implement the interface? It can provide the underlying batches through a different API.

Reply:
The problem is that, we don't really have a batch API in Spark SQL. We rely on type erasure and codegen hack to implement columnar scan. It's hardcoded in the engine: `SparkPlan#execute` returns `RDD[InternalRow]`.

if we have a RDD iterate over the rows in the batch, then whole stage codegen will break, as it iterates the input RDD and cast the record to `ColumnarBatch`.

Reply:
Where is the issue to fix this hack?

This seems like something that should never have happened. We can simply have an additional trait on a `RDD[InternalRow]` to fetch the underlying iterator of `ColumnarBatch`. The codegen path should check for that and use it.

Reply:
SGTM. Can we do it in followup? The type hack is there for years and the fix may be non-trivial. We also need to fix the file source side.

Reply:
> We can simply have an additional trait on a RDD[InternalRow] to fetch the underlying iterator of ColumnarBatch

BTW this still need to make a `RDD[ColumnarBatch]` to pretend itself as `RDD[InternalRow]`?

Reply:
We can fix it in a follow-up. Please open an issue for it and link to it from this PR.

Reply:
> BTW this still need to make a RDD[ColumnarBatch] to pretend itself as RDD[InternalRow]?

What I'm proposing is to make the RDD actually implement RDD[InternalRow] and allow accessing the batches directly through a different iterator.